File: api-service/src/modules/recommendation/trip-cost-config.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { supabase } from '../../config/supabase';

type AlgorithmRow = {
  id: string;
  name: string;
  is_active: boolean | null;
};

type ParameterRow = {
  parameter_name: string;
  default_value: number | string | null;
  current_value: number | string | null;
};

export type TransportMode = 'motorbike' | 'car' | 'taxi' | 'truck';
export type CapacityMode = 'motorbike' | 'car';

export interface TripCostConfig {
  childPriceRatio: number;
  /** VND/km fuel cost for ONE vehicle of that mode (self-drive). */
  transportCostPerKm: Record<TransportMode, number>;
  transportCostPerKmDefault: number;
  /** Seats per vehicle, used to compute how many vehicles a group needs. */
  vehicleCapacity: Record<CapacityMode, number>;
  /** VND/ngày/người lớn — mốc sàn dùng thay cho budget=0 khi retry lịch
   * trình bị infeasible vì ngân sách quá thấp, để plan retry vẫn hướng tới
   * rẻ nhất-khả-thi thay vì bỏ qua chi phí hoàn toàn. */
  minDailyBudgetFloor: number;
  /** VND/đêm/người — mốc sàn CHỖ Ở, tách riêng khỏi minDailyBudgetFloor
   * (đó là mốc sàn cho tham quan/ăn uống). Chỉ cộng thêm cho (số ngày - 1)
   * đêm — chuyến 1 ngày (đi về trong ngày) không qua đêm nên không cần mốc
   * này. Khớp FALLBACK_HOTEL_COST_PER_NIGHT/ROOM_CAPACITY bên ai-service
   * (planner.py) — dùng khi 1 khách sạn cụ thể chưa có giá riêng. */
  minHotelNightFloor: number;
}

// Same `ai_config` schema already used by TwoTowerConfigService and the
// admin algorithm-settings module — a new algorithm row, not a new
// integration. Shared between ai-service (Python) and this service so the
// child-price ratio, per-km fuel cost, and vehicle capacity can never drift
// apart between the two.
const ALGORITHM_NAME = 'trip_cost_config';
const ALGORITHM_DESCRIPTION =
  'Tham số chi phí chuyến đi dùng chung ai-service/api-service';

const PARAM_CHILD_PRICE_RATIO = 'child_price_ratio';
// Canonical mode keys shared with ai-service's trip_cost_config_service.py.
// TransportMode.ROAD (create-itinerary.dto.ts) maps onto "car" — the old
// SELF_DRIVE_TRANSPORT_COST_PER_KM already treated ROAD and CAR identically.
const TRANSPORT_MODES: readonly TransportMode[] = [
  'motorbike',
  'car',
  'taxi',
  'truck',
];
// Only motorbike/car are reachable in practice (dto.transportMode only maps
// to these two for self-drive); taxi/truck have no capacity concept here.
const CAPACITY_MODES: readonly CapacityMode[] = ['motorbike', 'car'];
const PARAM_TRANSPORT_PREFIX = 'transport_cost_per_km_';
const PARAM_TRANSPORT_DEFAULT = 'transport_cost_per_km_default';
const PARAM_CAPACITY_PREFIX = 'vehicle_capacity_';
const PARAM_MIN_DAILY_BUDGET_FLOOR = 'min_daily_budget_floor';
const PARAM_MIN_HOTEL_NIGHT_FLOOR = 'min_hotel_night_floor';

const DEFAULTS: Record<string, number> = {
  [PARAM_CHILD_PRICE_RATIO]: 0.7,
  // Ban đầu 300k (dựa theo chi phí quán ăn quan sát trong log debug 2026-07:
  // ~60k-180k/bữa × 2 bữa) — nhưng không tính tới vé vào cửa/hoạt động có
  // thể rất rẻ (VD Thảo Cầm Viên ~50-60k, Chợ Bến Thành đi dạo miễn phí),
  // khiến 1 chuyến thực tế ~100k/ngày vẫn hợp lý bị từ chối oan. Hạ xuống
  // 150k (2026-07-22, theo yêu cầu chủ dự án) — chừa biên an toàn hơn ví dụ
  // 100k phòng khi có ngày khác đắt hơn trong cùng chuyến, vẫn rẻ hơn hẳn
  // mức 300k cũ. Admin có thể chỉnh qua current_value không cần deploy lại,
  // giống mọi param khác trong file này.
  [PARAM_MIN_DAILY_BUDGET_FLOOR]: 150_000,
  // Khớp FALLBACK_HOTEL_COST_PER_NIGHT (400_000) / ROOM_CAPACITY (2) bên
  // ai-service/planner.py — giá phòng tối thiểu giả định khi 1 khách sạn cụ
  // thể chưa có estimated_cost riêng.
  [PARAM_MIN_HOTEL_NIGHT_FLOOR]: 200_000,
  // Cập nhật theo yêu cầu chủ dự án (2026-07-14): giá xăng xe/ô tô thực tế
  // đã tăng so với mức cấu hình cũ (450/520). Đây là giá cho MỘT xe — nhân
  // số xe cần dùng ở estimateSelfDriveTransportCost(), không nhân theo đầu
  // người trực tiếp.
  transport_cost_per_km_motorbike: 600,
  transport_cost_per_km_car: 1800,
  transport_cost_per_km_taxi: 18000,
  transport_cost_per_km_truck: 20000,
  [PARAM_TRANSPORT_DEFAULT]: 10000,
  vehicle_capacity_motorbike: 2,
  vehicle_capacity_car: 4,
};
const DESCRIPTIONS: Record<string, string> = {
  [PARAM_CHILD_PRICE_RATIO]:
    'Tỷ lệ giá trẻ em so với người lớn (vé/ăn uống và khách sạn)',
  transport_cost_per_km_motorbike: 'VND/km xăng xe máy tự lái (1 xe)',
  transport_cost_per_km_car: 'VND/km xăng ô tô tự lái (1 xe)',
  transport_cost_per_km_taxi: 'VND/km cho taxi',
  transport_cost_per_km_truck: 'VND/km cho xe tải',
  [PARAM_TRANSPORT_DEFAULT]: 'VND/km mặc định khi không rõ phương tiện',
  vehicle_capacity_motorbike: 'Số người tối đa mỗi xe máy',
  vehicle_capacity_car: 'Số người tối đa mỗi ô tô (xe 4 chỗ)',
  [PARAM_MIN_DAILY_BUDGET_FLOOR]:
    'VND/ngày/người lớn — mốc sàn chi phí tối thiểu dùng khi ngân sách nhập vào quá thấp để tạo lịch trình',
  [PARAM_MIN_HOTEL_NIGHT_FLOOR]:
    'VND/đêm/người — mốc sàn chỗ ở tối thiểu, chỉ áp dụng cho chuyến từ 2 ngày trở lên',
};
const MIN_MAX: Partial<Record<string, [number, number]>> = {
  [PARAM_CHILD_PRICE_RATIO]: [0, 1],
  vehicle_capacity_motorbike: [1, 10],
  vehicle_capacity_car: [1, 16],
  [PARAM_MIN_DAILY_BUDGET_FLOOR]: [1, 10_000_000],
  [PARAM_MIN_HOTEL_NIGHT_FLOOR]: [1, 10_000_000],
};

const DEFAULT_CONFIG: TripCostConfig = {
  childPriceRatio: DEFAULTS[PARAM_CHILD_PRICE_RATIO],
  transportCostPerKm: {
    motorbike: DEFAULTS.transport_cost_per_km_motorbike,
    car: DEFAULTS.transport_cost_per_km_car,
    taxi: DEFAULTS.transport_cost_per_km_taxi,
    truck: DEFAULTS.transport_cost_per_km_truck,
  },
  transportCostPerKmDefault: DEFAULTS[PARAM_TRANSPORT_DEFAULT],
  vehicleCapacity: {
    motorbike: DEFAULTS.vehicle_capacity_motorbike,
    car: DEFAULTS.vehicle_capacity_car,
  },
  minDailyBudgetFloor: DEFAULTS[PARAM_MIN_DAILY_BUDGET_FLOOR],
  minHotelNightFloor: DEFAULTS[PARAM_MIN_HOTEL_NIGHT_FLOOR],
};

@Injectable()
export class TripCostConfigService {
  private readonly logger = new Logger(TripCostConfigService.name);
  private cached: TripCostConfig | null = null;
  private cachedAt = 0;
  private readonly cacheTtlMs = 60_000;

  async getConfig(): Promise<TripCostConfig> {
    const now = Date.now();
    if (this.cached && now - this.cachedAt < this.cacheTtlMs) {
      return this.cached;
    }

    try {
      const config = await this.loadFromDb();
      this.cached = config;
      this.cachedAt = now;
      return config;
    } catch (err: any) {
      const message = err?.message ?? String(err);
      this.logger.warn(
        `Failed to load trip cost config, using defaults: ${message}`,
      );
      this.cached = DEFAULT_CONFIG;
      this.cachedAt = now;
      return DEFAULT_CONFIG;
    }
  }

  invalidateCache(): void {
    this.cached = null;
    this.cachedAt = 0;
  }

  /** Số xe cần dùng cho cả đoàn = ceil(số người / sức chứa mỗi xe). Xăng xe
   * là chi phí theo XE, không phải theo NGƯỜI. */
  vehiclesNeeded(headcount: number, mode: CapacityMode, config: TripCostConfig): number {
    const capacity = config.vehicleCapacity[mode] || DEFAULTS[`${PARAM_CAPACITY_PREFIX}${mode}`];
    return Math.max(1, Math.ceil(Math.max(1, headcount) / capacity));
  }

  private async ensureAlgorithm(): Promise<AlgorithmRow> {
    const { data, error } = await supabase
      .schema('ai_config')
      .from('algorithms')
      .select('id,name,is_active')
      .eq('name', ALGORITHM_NAME)
      .maybeSingle();
    if (error) {
      throw new Error(error.message);
    }
    if (data) {
      return data as AlgorithmRow;
    }

    const { data: inserted, error: insertError } = await supabase
      .schema('ai_config')
      .from('algorithms')
      .insert({
        name: ALGORITHM_NAME,
        description: ALGORITHM_DESCRIPTION,
        is_active: true,
      })
      .select('id,name,is_active')
      .single();
    if (insertError) {
      throw new Error(insertError.message);
    }
    return inserted as AlgorithmRow;
  }

  private async ensureParam(
    algorithmId: string,
    name: string,
  ): Promise<ParameterRow> {
    const { data, error } = await supabase
      .schema('ai_config')
      .from('algorithm_parameters')
      .select('parameter_name,default_value,current_value')
      .eq('algorithm_id', algorithmId)
      .eq('parameter_name', name)
      .maybeSingle();
    if (error) {
      throw new Error(error.message);
    }

    const codeDefault = DEFAULTS[name];

    if (data) {
      const dbDefault = Number(data.default_value);
      // Reconcile: if the code's default changed since this row was seeded
      // (e.g. corrected fuel-price estimate), refresh the DB default, and
      // only follow through on current_value if nobody customized it away
      // from the old default (preserve manual admin overrides).
      if (!Number.isFinite(dbDefault) || Math.abs(dbDefault - codeDefault) > 1e-9) {
        const currentValue = Number(data.current_value);
        const patchBody: Record<string, unknown> = { default_value: codeDefault };
        if (
          !Number.isFinite(currentValue) ||
          (Number.isFinite(dbDefault) && Math.abs(currentValue - dbDefault) < 1e-9)
        ) {
          patchBody.current_value = codeDefault;
        }
        const { data: patched, error: patchError } = await supabase
          .schema('ai_config')
          .from('algorithm_parameters')
          .update(patchBody)
          .eq('algorithm_id', algorithmId)
          .eq('parameter_name', name)
          .select('parameter_name,default_value,current_value')
          .single();
        if (!patchError && patched) {
          return patched as ParameterRow;
        }
      }
      return data as ParameterRow;
    }

    const minMax = MIN_MAX[name];
    const insertBody: Record<string, unknown> = {
      algorithm_id: algorithmId,
      parameter_name: name,
      default_value: codeDefault,
      current_value: codeDefault,
      description: DESCRIPTIONS[name] ?? '',
    };
    if (minMax) {
      insertBody.min_value = minMax[0];
      insertBody.max_value = minMax[1];
    }
    const { data: inserted, error: insertError } = await supabase
      .schema('ai_config')
      .from('algorithm_parameters')
      .insert(insertBody)
      .select('parameter_name,default_value,current_value')
      .single();
    if (insertError) {
      throw new Error(insertError.message);
    }
    return inserted as ParameterRow;
  }

  private async loadFromDb(): Promise<TripCostConfig> {
    const algorithm = await this.ensureAlgorithm();
    const paramNames = [
      PARAM_CHILD_PRICE_RATIO,
      PARAM_TRANSPORT_DEFAULT,
      PARAM_MIN_DAILY_BUDGET_FLOOR,
      PARAM_MIN_HOTEL_NIGHT_FLOOR,
      ...TRANSPORT_MODES.map((mode) => `${PARAM_TRANSPORT_PREFIX}${mode}`),
      ...CAPACITY_MODES.map((mode) => `${PARAM_CAPACITY_PREFIX}${mode}`),
    ];

    const params = new Map<string, ParameterRow>();
    for (const name of paramNames) {
      params.set(name, await this.ensureParam(algorithm.id, name));
    }

    const readNumber = (name: string): number => {
      const row = params.get(name);
      const value = Number(
        row?.current_value ?? row?.default_value ?? DEFAULTS[name],
      );
      return Number.isFinite(value) ? value : DEFAULTS[name];
    };

    return {
      childPriceRatio: readNumber(PARAM_CHILD_PRICE_RATIO),
      transportCostPerKm: {
        motorbike: readNumber(`${PARAM_TRANSPORT_PREFIX}motorbike`),
        car: readNumber(`${PARAM_TRANSPORT_PREFIX}car`),
        taxi: readNumber(`${PARAM_TRANSPORT_PREFIX}taxi`),
        truck: readNumber(`${PARAM_TRANSPORT_PREFIX}truck`),
      },
      transportCostPerKmDefault: readNumber(PARAM_TRANSPORT_DEFAULT),
      vehicleCapacity: {
        motorbike: Math.max(1, Math.round(readNumber(`${PARAM_CAPACITY_PREFIX}motorbike`))),
        car: Math.max(1, Math.round(readNumber(`${PARAM_CAPACITY_PREFIX}car`))),
      },
      minDailyBudgetFloor: readNumber(PARAM_MIN_DAILY_BUDGET_FLOOR),
      minHotelNightFloor: readNumber(PARAM_MIN_HOTEL_NIGHT_FLOOR),
    };
  }
}

File: backend/ai-service/app/services/trip_cost_config_service.py
"""Cấu hình chi phí chuyến đi dùng chung giữa ai-service và api-service.

Lưu trong cùng schema `ai_config` đã có sẵn cho hybrid recommender / two-tower
(xem `ai_config_service.py`), dưới một algorithm riêng tên `trip_cost_config`.
Mục đích: tránh lệch số giữa tỷ lệ giá trẻ em / bảng chi phí VND-km / sức chứa
xe của ai-service (Python) và api-service (NestJS) — cả hai đọc từ cùng các
dòng `algorithm_parameters` này thay vì mỗi bên tự hardcode một bản.

Chi phí xăng xe (transport_cost_per_km_*) là giá xăng thực tế cho **1 xe tự
lái** (không phải chi phí thuê xe dịch vụ). Vì 1 xe chở được nhiều người
(xe máy ~2 người, ô tô ~4 người), tổng chi phí của cả đoàn phải nhân theo
**số xe cần dùng** = ceil(số người / sức chứa mỗi xe) — xem
`vehicles_needed()`. Đừng nhân theo đầu người trực tiếp như child_price_ratio,
vì xăng xe là chi phí theo CHUYẾN/XE chứ không phải theo NGƯỜI.

Tự seed nếu thiếu, và tự đồng bộ lại nếu giá trị mặc định trong code đổi khác
với DB (giống `_ensure_candidate_count_param` trong `ai_config_service.py`) —
chỉ ghi đè `current_value` nếu nó đang bằng default cũ (chưa bị admin chỉnh
tay), để không mất tùy chỉnh thủ công.
"""

from __future__ import annotations

import math

import httpx

from app.core.config import settings
from app.core.logger import get_logger

logger = get_logger(__name__)

ALGO_NAME = "trip_cost_config"
ALGO_DESCRIPTION = "Tham so chi phi chuyen di dung chung ai-service/api-service"

PARAM_CHILD_PRICE_RATIO = "child_price_ratio"

# Canonical vehicle-mode keys shared with api-service's trip-cost-config.service.ts.
# api-service's TransportMode.ROAD maps onto "car" (SELF_DRIVE_TRANSPORT_COST_PER_KM
# already treated ROAD and CAR as the same rate before this consolidation).
TRANSPORT_MODES = ("motorbike", "car", "taxi", "truck")
# Only motorbike/car are reachable in practice (ai-service's travel_vehicle
# field only accepts "car"/"bike"); taxi/truck stay in the table for
# forward-compat but aren't user-selectable today.
CAPACITY_MODES = ("motorbike", "car")
PARAM_TRANSPORT_COST_PREFIX = "transport_cost_per_km_"
PARAM_TRANSPORT_COST_DEFAULT = "transport_cost_per_km_default"
PARAM_VEHICLE_CAPACITY_PREFIX = "vehicle_capacity_"

_DEFAULTS: dict[str, float] = {
    PARAM_CHILD_PRICE_RATIO: 0.7,
    # Giá xăng thực tế trung bình (7/2026): E5 RON92 ~19.730đ/l, tiêu hao
    # 2.0-2.5l/100km -> ~395-493đ/km, lấy ~450đ/km. RON95 ~20.410đ/l, ô tô
    # tiêu hao 7-10l/100km -> ~428-612đ/km, lấy ~520đ/km.
    "transport_cost_per_km_motorbike": 450,
    "transport_cost_per_km_car": 520,
    "transport_cost_per_km_taxi": 18_000,
    "transport_cost_per_km_truck": 20_000,
    PARAM_TRANSPORT_COST_DEFAULT: 10_000,
    "vehicle_capacity_motorbike": 2,
    "vehicle_capacity_car": 4,
}
_DESCRIPTIONS: dict[str, str] = {
    PARAM_CHILD_PRICE_RATIO: "Ty le gia tre em so voi nguoi lon (ve/an uong va khach san)",
    "transport_cost_per_km_motorbike": "VND/km xang xe may tu lai (1 xe)",
    "transport_cost_per_km_car": "VND/km xang o to tu lai (1 xe)",
    "transport_cost_per_km_taxi": "VND/km cho taxi",
    "transport_cost_per_km_truck": "VND/km cho xe tai",
    PARAM_TRANSPORT_COST_DEFAULT: "VND/km mac dinh khi khong ro phuong tien",
    "vehicle_capacity_motorbike": "So nguoi toi da moi xe may",
    "vehicle_capacity_car": "So nguoi toi da moi o to (xe 4 cho)",
}
_MIN_MAX: dict[str, tuple[float | None, float | None]] = {
    PARAM_CHILD_PRICE_RATIO: (0.0, 1.0),
    "vehicle_capacity_motorbike": (1, None),
    "vehicle_capacity_car": (1, None),
}

_TIMEOUT = 8.0
_CACHE_TTL_SECONDS = 60.0


class TripCostConfigError(RuntimeError):
    """Lỗi nghiệp vụ khi đọc/ghi cấu hình chi phí chuyến đi."""


class _PostgREST:
    """Client PostgREST tối giản (giống ai_config_service._PostgREST)."""

    def __init__(self, base_url: str, key: str, schema: str):
        if not base_url or not key:
            raise TripCostConfigError(
                "Chưa cấu hình SUPABASE_URL / SUPABASE_KEY cho ai-service (.env)"
            )
        self.endpoint = f"{base_url.rstrip('/')}/rest/v1"
        self.schema = schema
        self._auth = {"apikey": key, "Authorization": f"Bearer {key}"}

    def select(self, table: str, params: dict) -> list[dict]:
        headers = {**self._auth, "Accept-Profile": self.schema}
        r = httpx.get(
            f"{self.endpoint}/{table}", params=params, headers=headers, timeout=_TIMEOUT
        )
        r.raise_for_status()
        return r.json()

    def patch(self, table: str, params: dict, body: dict) -> list[dict]:
        headers = {
            **self._auth,
            "Content-Profile": self.schema,
            "Prefer": "return=representation",
        }
        r = httpx.patch(
            f"{self.endpoint}/{table}",
            params=params,
            json=body,
            headers=headers,
            timeout=_TIMEOUT,
        )
        r.raise_for_status()
        return r.json()

    def insert(self, table: str, body: dict) -> list[dict]:
        headers = {
            **self._auth,
            "Content-Profile": self.schema,
            "Prefer": "return=representation",
        }
        r = httpx.post(
            f"{self.endpoint}/{table}", json=body, headers=headers, timeout=_TIMEOUT
        )
        r.raise_for_status()
        return r.json()


_client: _PostgREST | None = None


def _db() -> _PostgREST:
    global _client
    if _client is None:
        _client = _PostgREST(
            settings.supabase_url, settings.supabase_key, settings.ai_config_schema
        )
    return _client


def _ensure_algorithm() -> dict:
    rows = _db().select(
        "algorithms", {"name": f"eq.{ALGO_NAME}", "select": "*", "limit": 1}
    )
    if rows:
        return rows[0]
    inserted = _db().insert(
        "algorithms",
        {"name": ALGO_NAME, "description": ALGO_DESCRIPTION, "is_active": True},
    )
    if inserted:
        return inserted[0]
    # Race with another process seeding concurrently: re-select.
    rows = _db().select(
        "algorithms", {"name": f"eq.{ALGO_NAME}", "select": "*", "limit": 1}
    )
    if not rows:
        raise TripCostConfigError(f"Không thể tạo/tìm algorithm '{ALGO_NAME}'")
    return rows[0]


def _ensure_param(algorithm_id: str, name: str) -> dict:
    rows = _db().select(
        "algorithm_parameters",
        {
            "algorithm_id": f"eq.{algorithm_id}",
            "parameter_name": f"eq.{name}",
            "select": "*",
            "limit": 1,
        },
    )
    code_default = _DEFAULTS[name]

    if rows:
        row = rows[0]
        db_default = row.get("default_value")
        # Reconcile: if the code's default changed since this row was seeded
        # (e.g. we corrected the fuel-price estimate), refresh the DB default,
        # and only follow through on current_value if nobody has customized
        # it away from the old default (preserve manual admin overrides).
        try:
            db_default_num = float(db_default) if db_default is not None else None
        except (TypeError, ValueError):
            db_default_num = None
        if db_default_num is None or abs(db_default_num - float(code_default)) > 1e-9:
            try:
                current_value_num = float(row.get("current_value"))
            except (TypeError, ValueError):
                current_value_num = None
            patch_body = {"default_value": code_default}
            if current_value_num is None or (
                db_default_num is not None
                and abs(current_value_num - db_default_num) < 1e-9
            ):
                patch_body["current_value"] = code_default
            patched = _db().patch(
                "algorithm_parameters",
                {
                    "algorithm_id": f"eq.{algorithm_id}",
                    "parameter_name": f"eq.{name}",
                },
                patch_body,
            )
            if patched:
                return patched[0]
        return row

    min_value, max_value = _MIN_MAX.get(name, (None, None))
    body = {
        "algorithm_id": algorithm_id,
        "parameter_name": name,
        "default_value": code_default,
        "current_value": code_default,
        "description": _DESCRIPTIONS.get(name, ""),
    }
    if min_value is not None:
        body["min_value"] = min_value
    if max_value is not None:
        body["max_value"] = max_value
    inserted = _db().insert("algorithm_parameters", body)
    if inserted:
        return inserted[0]
    rows = _db().select(
        "algorithm_parameters",
        {
            "algorithm_id": f"eq.{algorithm_id}",
            "parameter_name": f"eq.{name}",
            "select": "*",
            "limit": 1,
        },
    )
    if not rows:
        raise TripCostConfigError(f"Không thể tạo/tìm tham số '{name}'")
    return rows[0]


_cache: dict | None = None
_cache_loaded_at = 0.0


def get_trip_cost_config(force_refresh: bool = False) -> dict:
    """Trả về cấu hình dùng chung:
    {
      "child_price_ratio": float,
      "transport_cost_per_km": {mode: float},   # giá xăng/km cho 1 xe
      "transport_cost_per_km_default": float,
      "vehicle_capacity": {mode: int},           # so nguoi toi da moi xe
    }

    Cache trong tiến trình theo `_CACHE_TTL_SECONDS` để tránh gọi PostgREST mỗi
    lần lập lịch trình. Nếu DB lỗi/chưa cấu hình, fallback về `_DEFAULTS` và
    log cảnh báo — không được làm hỏng luồng lập lịch trình chính.
    """
    global _cache, _cache_loaded_at
    import time

    now = time.monotonic()
    if not force_refresh and _cache is not None and (now - _cache_loaded_at) < _CACHE_TTL_SECONDS:
        return _cache

    try:
        algo = _ensure_algorithm()
        param_names = (
            [PARAM_CHILD_PRICE_RATIO, PARAM_TRANSPORT_COST_DEFAULT]
            + [f"{PARAM_TRANSPORT_COST_PREFIX}{mode}" for mode in TRANSPORT_MODES]
            + [f"{PARAM_VEHICLE_CAPACITY_PREFIX}{mode}" for mode in CAPACITY_MODES]
        )
        params = {name: _ensure_param(algo["id"], name) for name in param_names}

        child_price_ratio = float(
            params[PARAM_CHILD_PRICE_RATIO].get("current_value")
            or _DEFAULTS[PARAM_CHILD_PRICE_RATIO]
        )
        transport_cost_per_km = {
            mode: float(
                params[f"{PARAM_TRANSPORT_COST_PREFIX}{mode}"].get("current_value")
                or _DEFAULTS[f"{PARAM_TRANSPORT_COST_PREFIX}{mode}"]
            )
            for mode in TRANSPORT_MODES
        }
        transport_cost_default = float(
            params[PARAM_TRANSPORT_COST_DEFAULT].get("current_value")
            or _DEFAULTS[PARAM_TRANSPORT_COST_DEFAULT]
        )
        vehicle_capacity = {
            mode: max(
                1,
                int(
                    float(
                        params[f"{PARAM_VEHICLE_CAPACITY_PREFIX}{mode}"].get(
                            "current_value"
                        )
                        or _DEFAULTS[f"{PARAM_VEHICLE_CAPACITY_PREFIX}{mode}"]
                    )
                ),
            )
            for mode in CAPACITY_MODES
        }

        config = {
            "child_price_ratio": child_price_ratio,
            "transport_cost_per_km": transport_cost_per_km,
            "transport_cost_per_km_default": transport_cost_default,
            "vehicle_capacity": vehicle_capacity,
        }
        _cache = config
        _cache_loaded_at = now
        return config
    except Exception as e:  # noqa: BLE001 - config must never break planning
        logger.warning(
            "Không tải được trip_cost_config từ DB, dùng giá trị mặc định: %s", e
        )
        fallback = {
            "child_price_ratio": _DEFAULTS[PARAM_CHILD_PRICE_RATIO],
            "transport_cost_per_km": {
                mode: _DEFAULTS[f"{PARAM_TRANSPORT_COST_PREFIX}{mode}"]
                for mode in TRANSPORT_MODES
            },
            "transport_cost_per_km_default": _DEFAULTS[PARAM_TRANSPORT_COST_DEFAULT],
            "vehicle_capacity": {
                mode: int(_DEFAULTS[f"{PARAM_VEHICLE_CAPACITY_PREFIX}{mode}"])
                for mode in CAPACITY_MODES
            },
        }
        if _cache is None:
            _cache = fallback
            _cache_loaded_at = now
        return fallback


def vehicles_needed(headcount: int, mode: str, config: dict | None = None) -> int:
    """Số xe cần dùng cho cả đoàn = ceil(số người / sức chứa mỗi xe).

    Xăng xe là chi phí theo XE, không phải theo NGƯỜI — 5 người đi xe máy cần
    3 xe (ceil(5/2)), không phải 1 xe tính x5 hay x1.
    """
    cfg = config or get_trip_cost_config()
    capacity = cfg.get("vehicle_capacity", {}).get(mode)
    if not capacity:
        capacity = int(_DEFAULTS.get(f"{PARAM_VEHICLE_CAPACITY_PREFIX}{mode}", 1))
    return max(1, math.ceil(max(1, int(headcount)) / capacity))


def sync_transport_cost_into_planner(force_refresh: bool = True) -> None:
    """Overwrite `planner.TRANSPORT_COST_PER_KM`/`TRANSPORT_COST_DEFAULT`/
    `VEHICLE_CAPACITY` in place with the shared config so ai-service and
    api-service can never diverge on VND/km or seats-per-vehicle.
    `planner.py`/`scheduler_v2.py` keep reading these same module-level names
    — no call-site changes needed, so a DB read failure here just means they
    keep whatever value was already loaded (module defaults on first boot).

    Call at process startup (see `app/main.py` lifespan, alongside
    `ai_config_service.sync_engine_from_db()`).
    """
    from app.services.itinerary import planner  # local import: avoid import cycle

    config = get_trip_cost_config(force_refresh=force_refresh)
    per_km = config["transport_cost_per_km"]
    capacity = config["vehicle_capacity"]
    # planner.py's own key is "bike" (not "motorbike") for the same vehicle —
    # historical naming in ai-service, kept as-is to avoid touching every
    # travel_vehicle="bike" call site across the codebase.
    planner.TRANSPORT_COST_PER_KM.clear()
    planner.TRANSPORT_COST_PER_KM.update(
        {
            "bike": per_km["motorbike"],
            "car": per_km["car"],
            "taxi": per_km["taxi"],
            "truck": per_km["truck"],
        }
    )
    planner.TRANSPORT_COST_DEFAULT = config["transport_cost_per_km_default"]
    planner.VEHICLE_CAPACITY.clear()
    planner.VEHICLE_CAPACITY.update(
        {
            "bike": capacity["motorbike"],
            "car": capacity["car"],
        }
    )
